# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/HTMLEditSubActionHandler.cpp
# Commit: dfd4ec691465
# Full Hash: dfd4ec69146560b6de52edbc4acdb49762daf38c
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-01-17 16:13:02
# Regressor Bug: 1807829
# File Overlap Count: 2
# Description:
#   Bug 1807829 - Make `HTMLEditor::CreateStyleForInsertText` use same logic as `HTMLEditor::SetInlinePropertiesAsSubAction` to apply new style r=m_kato
#   
#   The problem in Yahoo mail is, when you apply new font size to:
#   ```
#   <div><font size="7">{}<br></font></div>
# ==============================================================================

diff -r 0f2913c81558 -r dfd4ec691465 editor/libeditor/HTMLEditSubActionHandler.cpp
--- a/editor/libeditor/HTMLEditSubActionHandler.cpp	Mon Jan 16 23:35:34 2023 +0000
+++ b/editor/libeditor/HTMLEditSubActionHandler.cpp	Mon Jan 16 23:42:27 2023 +0000
@@ -1094,7 +1094,7 @@
 
   // for every property that is set, insert a new inline style node
   Result<EditorDOMPoint, nsresult> setStyleResult =
-      CreateStyleForInsertText(pointToInsert);
+      CreateStyleForInsertText(pointToInsert, *editingHost);
   if (MOZ_UNLIKELY(setStyleResult.isErr())) {
     NS_WARNING("HTMLEditor::CreateStyleForInsertText() failed");
     return setStyleResult.propagateErr();
@@ -2258,7 +2258,7 @@
   //       should be merged when we fix bug 92921.
 
   Result<EditorDOMPoint, nsresult> setStyleResult =
-      CreateStyleForInsertText(aPointToBreak);
+      CreateStyleForInsertText(aPointToBreak, aEditingHost);
   if (MOZ_UNLIKELY(setStyleResult.isErr())) {
     NS_WARNING("HTMLEditor::CreateStyleForInsertText() failed");
     return setStyleResult.propagateErr();
@@ -6078,7 +6078,7 @@
 }
 
 Result<EditorDOMPoint, nsresult> HTMLEditor::CreateStyleForInsertText(
-    const EditorDOMPoint& aPointToInsertText) {
+    const EditorDOMPoint& aPointToInsertText, const Element& aEditingHost) {
   MOZ_ASSERT(IsEditActionDataAvailable());
   MOZ_ASSERT(aPointToInsertText.IsSetAndValid());
   MOZ_ASSERT(mPendingStylesToApplyToNewContent);
@@ -6116,9 +6116,17 @@
   // then process setting any styles
   const int32_t relFontSize =
       mPendingStylesToApplyToNewContent->TakeRelativeFontSize();
-  pendingStyle = mPendingStylesToApplyToNewContent->TakePreservedStyle();
-
-  if (pendingStyle || relFontSize) {
+  AutoTArray<EditorInlineStyleAndValue, 32> stylesToSet;
+  mPendingStylesToApplyToNewContent->TakeAllPreservedStyles(stylesToSet);
+  if (stylesToSet.IsEmpty() && !relFontSize) {
+    return pointToPutCaret;
+  }
+
+  // We're in chrome, e.g., the email composer of Thunderbird, and there is
+  // relative font size changes, we need to keep using legacy path until we port
+  // IncrementOrDecrementFontSizeAsSubAction() to work with
+  // AutoInlineStyleSetter.
+  if (relFontSize) {
     // we have at least one style to add; make a new text node to insert style
     // nodes above.
     EditorDOMPoint pointToInsertTextNode(pointToPutCaret);
@@ -6164,34 +6172,26 @@
     insertNewTextNodeResult.inspect().IgnoreCaretPointSuggestion();
     pointToPutCaret.Set(newEmptyTextNode, 0u);
 
-    if (relFontSize) {
-      HTMLEditor::FontSize incrementOrDecrement =
-          relFontSize > 0 ? HTMLEditor::FontSize::incr
-                          : HTMLEditor::FontSize::decr;
-      for ([[maybe_unused]] uint32_t j : IntegerRange(Abs(relFontSize))) {
-        Result<CreateElementResult, nsresult>
-            wrapTextInBigOrSmallElementResult = SetFontSizeOnTextNode(
-                *newEmptyTextNode, 0, UINT32_MAX, incrementOrDecrement);
-        if (MOZ_UNLIKELY(wrapTextInBigOrSmallElementResult.isErr())) {
-          NS_WARNING("HTMLEditor::SetFontSizeOnTextNode() failed");
-          return wrapTextInBigOrSmallElementResult.propagateErr();
-        }
-        // We don't need to update here because we'll suggest caret position
-        // which is computed above.
-        MOZ_ASSERT(pointToPutCaret.IsSet());
-        wrapTextInBigOrSmallElementResult.inspect()
-            .IgnoreCaretPointSuggestion();
-      }
-    }
-
-    while (pendingStyle) {
-      AutoInlineStyleSetter inlineStyleSetter(
-          pendingStyle->GetAttribute()
-              ? EditorInlineStyleAndValue(
-                    *pendingStyle->GetTag(), *pendingStyle->GetAttribute(),
-                    pendingStyle->AttributeValueOrCSSValueRef())
-              : EditorInlineStyleAndValue(*pendingStyle->GetTag()));
-      // MOZ_KnownLive(...ContainerAs<nsIContent>()) because pointToPutCaret()
+    HTMLEditor::FontSize incrementOrDecrement =
+        relFontSize > 0 ? HTMLEditor::FontSize::incr
+                        : HTMLEditor::FontSize::decr;
+    for ([[maybe_unused]] uint32_t j : IntegerRange(Abs(relFontSize))) {
+      Result<CreateElementResult, nsresult> wrapTextInBigOrSmallElementResult =
+          SetFontSizeOnTextNode(*newEmptyTextNode, 0, UINT32_MAX,
+                                incrementOrDecrement);
+      if (MOZ_UNLIKELY(wrapTextInBigOrSmallElementResult.isErr())) {
+        NS_WARNING("HTMLEditor::SetFontSizeOnTextNode() failed");
+        return wrapTextInBigOrSmallElementResult.propagateErr();
+      }
+      // We don't need to update here because we'll suggest caret position
+      // which is computed above.
+      MOZ_ASSERT(pointToPutCaret.IsSet());
+      wrapTextInBigOrSmallElementResult.inspect().IgnoreCaretPointSuggestion();
+    }
+
+    for (const EditorInlineStyleAndValue& styleToSet : stylesToSet) {
+      AutoInlineStyleSetter inlineStyleSetter(styleToSet);
+      // MOZ_KnownLive(...ContainerAs<nsIContent>()) because pointToPutCaret
       // grabs the result.
       Result<CaretPoint, nsresult> setStyleResult =
           inlineStyleSetter.ApplyStyleToNodeOrChildrenAndRemoveNestedSameStyle(
@@ -6204,10 +6204,43 @@
       // is computed above.
       MOZ_ASSERT(pointToPutCaret.IsSet());
       setStyleResult.unwrap().IgnoreCaretPointSuggestion();
-      pendingStyle = mPendingStylesToApplyToNewContent->TakePreservedStyle();
-    }
-  }
-
+    }
+    return pointToPutCaret;
+  }
+
+  // If we have preserved commands except relative font style changes, we can
+  // use inline style setting code which reuse ancestors better.
+  AutoRangeArray ranges(pointToPutCaret);
+  if (MOZ_UNLIKELY(ranges.Ranges().IsEmpty())) {
+    NS_WARNING("AutoRangeArray::AutoRangeArray() failed");
+    return Err(NS_ERROR_FAILURE);
+  }
+  nsresult rv =
+      SetInlinePropertiesAroundRanges(ranges, stylesToSet, aEditingHost);
+  if (NS_FAILED(rv)) {
+    NS_WARNING("HTMLEditor::SetInlinePropertiesAroundRanges() failed");
+    return Err(rv);
+  }
+  if (NS_WARN_IF(ranges.Ranges().IsEmpty())) {
+    return Err(NS_ERROR_FAILURE);
+  }
+  // Now `ranges` selects new styled contents and the range may not be
+  // collapsed.  We should use the deepest editable start point of the range
+  // to insert text.
+  nsINode* container = ranges.FirstRangeRef()->GetStartContainer();
+  if (MOZ_UNLIKELY(!container->IsContent())) {
+    container = ranges.FirstRangeRef()->GetChildAtStartOffset();
+    if (MOZ_UNLIKELY(!container)) {
+      NS_WARNING("How did we get lost insertion point?");
+      return Err(NS_ERROR_FAILURE);
+    }
+  }
+  pointToPutCaret =
+      HTMLEditUtils::GetDeepestEditableStartPointOf<EditorDOMPoint>(
+          *container->AsContent());
+  if (NS_WARN_IF(!pointToPutCaret.IsSet())) {
+    return Err(NS_ERROR_FAILURE);
+  }
   return pointToPutCaret;
 }
 